### PR TITLE
ISPN-13165 AsyncOfflineTest.testReset random failures

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-13165

### CODE REVIEW

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 1]
```diff
@@ -148,7 +148,10 @@ public void trackUpdatedKey(int segment, Object key, Object lockOwner) {
       if (log.isTraceEnabled()) {
          log.tracef("Tracking key for %s: %s", lockOwner, key);
       }
-      updatedKeys.put(key, new State(segment, key, lockOwner));
+      State old = updatedKeys.put(key, new State(segment, key, lockOwner));
+      if (old != null) {
+         old.discard(); //avoid sending the cleanup command to the cluster members
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:153]: Nitpicking, but IMO the comment is too long, it should be on its own line (above)

File: core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java [hunk 2]
```diff
@@ -116,13 +120,23 @@ public void testReset(Method method) {
       List<DiscardInboundHandler> handlers = replaceSFOInboundHandler();
       handlers.forEach(h -> h.discard = true);
 
-      //SFO request should timeout
+      //SFO request should timeout (timeout=1 sec) and it retries (max retries = 6)
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java:123]: Shouldn't the test configure a lower number of retries to make things faster?

**Review by danberindei — CHANGES_REQUESTED** [review]
It looks like the problem wasn't in the test, what about creating a new JIRA or editing this one to describe what was wrong?

**Comment by pruivo** [comment]
@danberindei the problem was the test. I just noticed some IracCleanupKeyCommands in the logs that weren't required to be sent, i.e they will be discarded in the backup owners because they were from outdated writes. 

The problem was a TimeoutException being triggered after line 128 from retries. The for loop ensures everything is flushed.

**Comment by danberindei** [comment]
@pruivo so the problem is that IRAC kept retrying for > 30s, and `eventually()` gave up? How does your fix deal with that?

File: core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java (hunk 2 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/xsite/offline/AsyncOfflineTest.java:123]: no. the test only waits for the first failure and it should never be reached

**Comment by pruivo** [comment]
No. 
The test induces a failure by discarding the command in the receiving site (`discard=true`, line 121); this will throw a `TimeoutException`.
After the `TimeoutException`, the failure count is incremented and `eventuallyEquals(1, lonStatus::getFailureCount);` completes.
However, a retry happens after the `TimeoutException` which I didn't consider and, if unlucky, it can be discarded as well. If it is discarded, the following event order may happen (randomly)
1. `lonCache.put(key, "value3");` is sent and acked. The ack will reset the failure count to 0
2. `TimeoutException` is thrown, failure counter is increment
3. The last assertion `eventuallyEquals(0, lonStatus::getFailureCount);` will fail.

The for-loop added will protect the `lonCache.put(key, "value3");` by making sure the key is really flushed (i.e. if `TimeoutException` throws again, the next retry will succeed).

**Comment by pruivo** [comment]
PR updated

**Comment by ryanemerson** [comment]
Thanks @pruivo

**Pull request merged**